backend/danswer/background/connector_deletion.py

```diff
@@ -0,0 +1,296 @@
+        # if a document store entry is only indexed by this connector_credential_pair, delete it
+        num_docs_deleted = 0
+        chunks_to_delete = get_chunks_with_single_connector_credential_pair(
+            db_session=db_session,
+            connector_id=connector_id,
+            credential_id=credential_id,
+        )
+        if chunks_to_delete:
+            document_ids: set[str] = set()
+            vector_chunk_ids_to_delete: list[str] = []
+            keyword_chunk_ids_to_delete: list[str] = []
+            for chunk in chunks_to_delete:
+                document_ids.add(chunk.document_id)
+                if chunk.document_store_type == StoreType.KEYWORD:
+                    keyword_chunk_ids_to_delete.append(chunk.id)
+                else:
+                    vector_chunk_ids_to_delete.append(chunk.id)
+
+            vector_index.delete(ids=vector_chunk_ids_to_delete)
+            keyword_index.delete(ids=keyword_chunk_ids_to_delete)
+            # removes all `Chunk`, `DocumentByConnectorCredentialPair`, and `Document`
+            # rows from the DB
+            delete_documents_complete(
+                db_session=db_session,
+                document_ids=list(document_ids),
+            )
+            num_docs_deleted += len(document_ids)
+        return num_docs_deleted
+
+    num_docs_deleted = _delete_singly_indexed_docs()
+
+    def _update_multi_indexed_docs() -> None:
+        # if a document is indexed by multiple connector_credential_pairs, we should
+        # update it's access rather than outright delete it
+        document_by_connector_credential_pairs_to_update = (
+            get_document_by_connector_credential_pairs_indexed_by_multiple(
+                db_session=db_session,
+                connector_id=connector_id,
+                credential_id=credential_id,
+            )
+        )
+
+        def _get_user(
+            credential: Credential,
+        ) -> str:
+            if credential.public_doc:
+                return PUBLIC_DOC_PAT
+
+            return str(credential.user.id)
+
+        # find out which documents need to be updated and what their new allowed_users
+        # should be. This is a bit slow as it requires looping through all the documents
+        to_be_deleted_user = _get_user(deletion_attempt.credential)
+        document_ids_not_needing_update: set[str] = set()
+        document_id_to_allowed_users: dict[str, list[str]] = defaultdict(list)
+        for (
+            document_by_connector_credential_pair
+        ) in document_by_connector_credential_pairs_to_update:
+            document_id = document_by_connector_credential_pair.id
+            user = _get_user(document_by_connector_credential_pair.credential)
+            document_id_to_allowed_users[document_id].append(user)
+
+            # if there's another connector / credential pair which has indexed this
+            # document with the same access, we don't need to update it since removing
```

Comment:
clearer to say with credential belonging to the same user as the credential being deleted, but whatever eh? Also bit iffy if a credential belongs to multiple users at some future point, but ya nvm its good for now